lang: add "generated" symbol kind

alaviss opened the pull request:
## Summary
* Adds  `skGenerated`  symbol kind and diagnostics for
used-but-undeclared symbols.
*  `macros.genSym`  no longer takes a  `kind`  parameter and instead
will convert to the correct type on declaration.

## Details
Previously, generated symbols are specialized directly into their target
kind at generation site. This however throws a wrench into the
compiler's assumption, where specialized symbols are considered to be
well-formed and can be used as is.

This PR introduces a new  `skGenerated`  symbol type, which is used by
all generated symbols until their specialization (i.e. a declaration
statement happened for the symbol). This allows symbol usage sites
within the compiler to catch use-before-declaration errors and emit
appropriate diagnostics.

The automatic specialization also provided a UX improvement for macros
users, where  `genSym`  no longer requires a  `kind`  to be specified
and can automatically morph to the correct type once appropriate ASTs
are emitted. The old form of  `genSym`  has been marked as deprecated.

Fixes https://github.com/nim-works/nimskull/issues/1539

alaviss committed:
lang: add "generated" symbol kind

Previously, generated symbols are specialized directly into their target
kind at generation site. This however throw a wrench into the compiler's
assumption, where specialized symbols are considered to be well-formed
and can be used as is.

This commit introduces a new `skGenerated` symbol type, which is used by
all generated symbols until their specialization (i.e. a declaration
statement happened for the symbol). This allows symbol usage sites
within the compiler to catch use-before-declaration errors and emit
appropriate diagnostics.

The automatic specialization also provided an UX improvement for macros
users, where `genSym` no longer requires a `kind` to be specified and
can automatically morph to the correct type once appropriate AST is
emitted.

saem commented on tests/errmsgs/tundeclared_symbol.nim:
```diff
@@ -0,0 +1,14 @@
+discard """
+  cmd: '''nim check --msgFormat:sexp --filenames:canonical --hints:off $file'''
+  action: reject
+  nimoutFormat: sexp
```
Please don't use `sexp`, that whole thing needs to be removed.

alaviss committed:
tundeclared_symbol: switch to nimout matching

alaviss committed:
macros: remove default `ident` parameter for old gensym

Otherwise this causes `genSym()` to be ambiguous. The older variant is
still required for bootstrap and is moved to another branch.

alaviss committed:
std/tasks: convert to new genSym

I wanted to avoid this, but its usage of genSym cause ambiguous calls

alaviss committed:
semtempl: extend isTemplParam to handle skGenerated

alaviss committed:
ast_types: reorder skGenerated to before skPackage

This allows the gap in macros.nim to be removed

alaviss committed:
tastspec: adapt to current gensym

alaviss committed:
semtempl: undo skGenerated-related changes

semtempl uses specialized genSym to aid in several of its decisions,
including overloading resolution within templates.

alaviss committed:
tdefer_malformed_many_children: fix line number for macros.nim

zerbina commented on compiler/vm/vmgen.nim:
```diff
@@ -2003,7 +2003,13 @@ proc genMagic(c: var TCtx; n: CgNode; dest: var TDest; m: TMagic) =
     c.gABC(n, opcCallSite, dest)
-  of mNGenSym: genBinaryABC(c, n, dest, opcGenSym)
+  of mNGenSym:
+    if n.len == 2:
+      # Unary genSym
+      genUnaryABC(c, n, dest, opcGenSym)
+    else:
+      # Deprecated binary form
+      genBinaryABC(c, n, dest, opcGenSym)
```
```suggestion
    prepare(c, dest, n.typ)
    let arg =
      if n.len == 2: # Unary genSym
        c.genx(n[1])
      else: # Deprecated binary form
        c.freeTemp(c.genx(n[1])) # eval the first argument but discard the result
        c.genx(n[2])
    c.gABC(n, opcGenSym, dest, arg)
    c.freeTemp(arg)
```
A little more code, but all legacy compatibility is implemented in the code generator this way.

alaviss replied:
Would it be better to just cut a new csources after this, then we can remove this compat generation code and switch to templates?

Actually I'm already `when`-ing the thing, so switching to templates might not be too big of a deal.

zerbina replied:
Given that `genSym` is a magic, meaning that using it as a first-class procedure is not possible anyway, using a template + `.line` would be better, yea.

zerbina commented on compiler/vm/vm.nim:
```diff
@@ -2707,15 +2707,19 @@ proc rawExecute(c: var TCtx, t: var VmThread, pc: var int): YieldReason =
+          rb
+
+      checkHandle(regs[rh])
+      assert regs[rh].handle.typ.kind == akString
```
```suggestion
      checkHandle(regs[rb])
      assert regs[rb].handle.typ.kind == akString
```
A follow-up to the above comment.

zerbina commented on compiler/vm/vm.nim:
```diff
@@ -2707,15 +2707,19 @@ proc rawExecute(c: var TCtx, t: var VmThread, pc: var int): YieldReason =
-        "request to create symbol of invalid kind")
-      var sym = newSym(k.TSymKind, getIdent(c.cache, name), nextSymId c.idgen, c.module.owner, c.debug[pc])
+      let name = if regs[rh].strVal.len == 0: ":tmp"
+                 else: $regs[rh].strVal
```
```suggestion
      let name = if regs[rb].strVal.len == 0: ":tmp"
                 else: $regs[rb].strVal
```

zerbina commented on tests/errmsgs/tundeclared_symbol.nim:
```diff
@@ -0,0 +1,15 @@
+discard """
+  cmd: '''nim check --hints:off $file'''
+  action: reject
+  nimout: '''
+tundeclared_symbol.nim(12, 17) Error: symbol used before declaration: 'c' [generated in tundeclared_symbol.nim(12, 17)]
+'''
```
```suggestion
  errormsg: "symbol used before declaration: 'c' [generated in tundeclared_symbol.nim(9, 17)]"
  line: 9
```
This is more in line with the other tests from `errmsgs`. I've also modified the line numbers, so applying the change should just work.

zerbina commented on tests/lang_callable/macros/tsymbol_must_match.nim:
```diff
@@ -1,68 +1,120 @@
+proc nameExpr(n: NimNode): NimNode =
+  # Extracts the expression containing the replacement name
+  case n.kind
+  of nnkStmtList, nnkStmtListExpr, nnkForStmt:
+    nameExpr n[^1]
+  of nnkConstSection, nnkConstDef:
+    nameExpr n[0]
+  of nnkDiscardStmt:
+    nameExpr n[0]
+  of RoutineNodes:
+    n.name
+  of AtomicNodes:
+    n
```
```suggestion
  of nnkStmtList:
    nameExpr n[^1]
  of nkLetSection, nkVarSection, nkConstSection:
    n[0][0]
  of RoutineNodes:
    n.name
```
I would suggest just always using routines and let/var statements as the symbol suppliers. The only reason I sometimes used `nskForVar` was because it cost effectively nothing.

zerbina commented on tests/lang_callable/macros/tsymbol_must_match.nim:
```diff
@@ -1,68 +1,120 @@
 
-replaceName(nskVar):
+replaceName:
+  for iter in [0]:
+    discard iter
```
```suggestion
  var iter = 0
```
A follow-up to the above suggestion.

zerbina commented on tests/lang_callable/macros/tsymbol_must_match.nim:
```diff
@@ -1,68 +1,120 @@
 
-replaceName(nskForVar):
+replaceName:
+  for conv in [0]:
+    discard conv
```
```suggestion
  proc conv() = discard
```
A follow-up to the above suggestion.

zerbina commented on tests/lang_callable/macros/tsymbol_must_match.nim:
```diff
@@ -1,68 +1,120 @@
-replaceName(nskForVar):
-  method meth() = #[tt.Error
+replaceName:
+  for meth in [0]:
+    discard meth
```
```suggestion
  proc meth() = discard
```

zerbina commented on tests/lang_callable/macros/tsymbol_must_match.nim:
```diff
@@ -1,68 +1,120 @@
-               ^ (SemSymbolKindMismatch)]#
-  result = newVarStmt(b, newLit 1)
+replaceName:
+  let a = 0
+  a
```
```suggestion
  let a = 0
```
A follow-up to the above suggestion.

zerbina commented on tests/lang_callable/macros/tsymbol_must_match.nim:
```diff
@@ -1,68 +1,120 @@
 
-t1()
+replaceName:
+  proc c = discard
+  c
```
```suggestion
  proc c() = discard
```
So that it's stylistically in line with the other test cases.

zerbina commented on tests/lang_callable/macros/tsymbol_must_match.nim:
```diff
@@ -1,68 +1,120 @@
+       ^ (SemSymbolKindMismatch)]#
+
+replaceName:
+  func T = discard
+  T
```
```suggestion
  func T() = discard
```
Same here.

zerbina reviewed (commented):
Looks great! I've left some minor suggestions regarding the tests, plus a suggestion for handling all `genSym` compatibility in `vmgen`.

For the PR description, could you add to the summary a mention regarding the `kind` parameter of `genSym` being ignored now and fully optional?

zerbina commented:
@alaviss: Regarding your question about whether the first-pass for templates should use `skGenerated`, my answer would be no. The name `newGenSym` is somewhat of a misnomer, as while the symbols are "generated", they're not in the same way as it happens for macros. `skGenerated` makes sense for user-generated symbols, but for templates, the symbols and their creation are controlled by the compiler.

alaviss committed:
tundeclared_symbol: switch to errmsg matching

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

alaviss committed:
tsymbol_must_match: simplify test cases and some formatting

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

alaviss committed:
macros: use compatibility templates instead of procs

This lets us remove the compatibility code from vm and vmgen, keeping
them in stdlib instead.

alaviss committed:
tsymbol_must_match: fix enum names

zerbina commented on lib/core/macros.nim:
```diff
@@ -432,10 +432,33 @@ proc bindSym*(ident: string | NimNode, rule: BindSymRule = brClosed): NimNode {.
+    ##
+    ## This is a compatibility alias for `genSym <#genSym,string>`_, the `kind`
+    ## parameter is ignored.
+    genSym(ident)
```
```suggestion
    {.line.}:
      discard kind 
      genSym(ident)
```
Even if the routine is deprecated, I still think it makes sense to keep its behaviour roughly as it was, especially if it as simple as it is here.

zerbina commented on lib/core/macros.nim:
```diff
@@ -432,10 +432,33 @@ proc bindSym*(ident: string | NimNode, rule: BindSymRule = brClosed): NimNode {.
+    ##
+    ## This is a compatibility alias for `genSym <#genSym,string>`_, the `kind`
+    ## parameter is ignored.
+    genSym()
```
```suggestion
    {.line.}:
      discard kind
      genSym()
```

zerbina commented on tests/lang_stmts/defer/tdefer_malformed_many_children.nim:
```diff
@@ -2,7 +2,7 @@ discard """
   errormsg: "illformed AST"
   file: "macros.nim"
-  line: 647
+  line: 670
```
```suggestion
  line: 674
```
Adjusts the line number to still be correct after the two previous changes.

zerbina reviewed (approved):
One minor suggestion for keeping backwards compatibility for `genSym`, as it's super easy to do in this case.

alaviss committed:
macros: evaluate `kind` in compatibility gensym templates

This makes sure that expressions in `kind` are still executed alongside their side-effects (if any)

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

zerbina commented:
/merge
